File: agent/tools/policy_helpers.go
package tools

import (
	"fmt"

	"github.com/vultisig/vultisig-win/agent/shared"
)

func enrichPolicyFields(policy map[string]any, config map[string]any) {
	if config == nil {
		return
	}

	if assetObj, ok := config["asset"].(map[string]any); ok {
		chain, token, ticker := resolveAssetObj(assetObj)
		if ticker != "" {
			policy["from_asset"] = ticker
			policy["fromAsset"] = ticker
			policy["from_chain"] = chain
		}
		if recipients, ok := config["recipients"].([]any); ok && len(recipients) > 0 {
			if r, ok := recipients[0].(map[string]any); ok {
				if addr, ok := r["toAddress"]; ok {
					policy["to_address"] = fmt.Sprintf("%v", addr)
				}
				if amt, ok := r["amount"]; ok {
					policy["amount"] = shared.FormatHumanAmount(fmt.Sprintf("%v", amt), chain, token)
				}
			}
		}
	} else {
		fromChain, fromToken, fromTicker := resolveConfigAsset(config, "from")
		if fromTicker != "" {
			policy["from_asset"] = fromTicker
			policy["fromAsset"] = fromTicker
			policy["from_chain"] = fromChain
		}

		toChain, _, toTicker := resolveConfigAsset(config, "to")
		if toTicker != "" {
			policy["to_asset"] = toTicker
			policy["toAsset"] = toTicker
			policy["to_chain"] = toChain
		}

		if rawAmount, ok := config["fromAmount"]; ok {
			amountStr := fmt.Sprintf("%v", rawAmount)
			policy["amount"] = shared.FormatHumanAmount(amountStr, fromChain, fromToken)
		}
	}

	if freq, ok := config["frequency"]; ok {
		policy["frequency"] = fmt.Sprintf("%v", freq)
		policy["schedule"] = fmt.Sprintf("%v", freq)
	}
}

func resolveAssetObj(obj map[string]any) (chain, token, ticker string) {
	chain = fmt.Sprintf("%v", obj["chain"])
	if t, ok := obj["token"]; ok && t != nil {
		token = fmt.Sprintf("%v", t)
	}
	ticker = shared.ResolveTickerByChainAndToken(chain, token)
	return chain, token, ticker
}

func resolveConfigAsset(config map[string]any, field string) (chain, token, ticker string) {
	raw, ok := config[field]
	if !ok {
		return "", "", ""
	}
	obj, ok := raw.(map[string]any)
	if !ok {
		return "", "", ""
	}
	return resolveAssetObj(obj)
}

File: core/ui/agent/tools/shared/policyHelpers.ts
import {
  formatHumanAmount,
  resolveTickerByChainAndToken,
} from './assetResolution'

function resolveAssetObj(obj: Record<string, unknown>): {
  chain: string
  token: string
  ticker: string
} {
  const chain = String(obj.chain ?? '')
  let token = ''
  if (obj.token != null) {
    token = String(obj.token)
  }
  const ticker = resolveTickerByChainAndToken(chain, token)
  return { chain, token, ticker }
}

function resolveConfigAsset(
  config: Record<string, unknown>,
  field: string
): { chain: string; token: string; ticker: string } {
  const raw = config[field]
  if (!raw || typeof raw !== 'object')
    return { chain: '', token: '', ticker: '' }
  return resolveAssetObj(raw as Record<string, unknown>)
}

export function enrichPolicyFields(
  policy: Record<string, unknown>,
  config: Record<string, unknown>
): void {
  if (!config) return

  const assetObj = config.asset
  if (assetObj && typeof assetObj === 'object') {
    const { chain, token, ticker } = resolveAssetObj(
      assetObj as Record<string, unknown>
    )
    if (ticker) {
      policy.from_asset = ticker
      policy.fromAsset = ticker
      policy.from_chain = chain
    }
    const recipients = config.recipients
    if (Array.isArray(recipients) && recipients.length > 0) {
      const r = recipients[0] as Record<string, unknown>
      if (r.toAddress != null) {
        policy.to_address = String(r.toAddress)
      }
      if (r.amount != null) {
        policy.amount = formatHumanAmount(String(r.amount), chain, token)
      }
    }
  } else {
    const from = resolveConfigAsset(config, 'from')
    if (from.ticker) {
      policy.from_asset = from.ticker
      policy.fromAsset = from.ticker
      policy.from_chain = from.chain
    }

    const to = resolveConfigAsset(config, 'to')
    if (to.ticker) {
      policy.to_asset = to.ticker
      policy.toAsset = to.ticker
      policy.to_chain = to.chain
    }

    if (config.fromAmount != null) {
      policy.amount = formatHumanAmount(
        String(config.fromAmount),
        from.chain,
        from.token
      )
    }
  }

  if (config.frequency != null) {
    policy.frequency = String(config.frequency)
    policy.schedule = String(config.frequency)
  }
}

export function computeNextExecution(
  createdAt: string,
  frequency: string
): string {
  const created = new Date(createdAt)
  if (isNaN(created.getTime())) return ''

  const now = new Date()

  if (frequency === 'monthly') {
    const next = new Date(created)
    while (next <= now) {
      next.setUTCMonth(next.getUTCMonth() + 1)
    }
    return next.toISOString()
  }

  let intervalMs: number
  switch (frequency) {
    case 'hourly':
      intervalMs = 60 * 60 * 1000
      break
    case 'daily':
      intervalMs = 24 * 60 * 60 * 1000
      break
    case 'weekly':
      intervalMs = 7 * 24 * 60 * 60 * 1000
      break
    default:
      return ''
  }

  const elapsed = now.getTime() - created.getTime()
  if (elapsed < 0) return created.toISOString()

  const periods = Math.floor(elapsed / intervalMs) + 1
  const next = new Date(created.getTime() + periods * intervalMs)
  return next.toISOString()
}
